### PR TITLE
Added the option to return value if a form failed

### PR DESCRIPTION
**Pull request by Juje**
Added the `setValidatorReturnValueOnFail()` method to the `Forms\Form` class so that it can manipulate what should happen if a form validation fails.

In other words a developer can call `$form->setValidatorReturnValueOnFail()` to make sure that data failed data will be put back into the form element.

### CODE REVIEW

**Comment by Juje** [comment]
@jlambe Any update when this can be merged?

**Comment by jlambe** [comment]
@Juje Sorry, it went off my radar... but can you add tests for this feature before I can merge ?

**Comment by Juje** [comment]
> @Juje Sorry, it went off my radar... but can you add tests for this feature before I can merge ?

Sure no problem

File: src/Forms/Form.php [hunk 1]
```diff
@@ -183,6 +183,13 @@ class Form extends HtmlBuilder implements FormInterface, FieldTypeInterface
      * @var array
      */
     private $data = [];
+    
+    /**
+     * If the validator should return the value on fail or not.
+     *
+     * @var bool
+     */
+    private $validatorReturnValueOnFail = false;
```
Comment (jlambe) [inline src/Forms/Form.php:192]:
```suggestion

```

I know that a dedicated property and a method could be better but in order to follow existing form API, let's introduce a form option directly and leverage the `$allowedOptions` and `$defaultOptions` properties of the Form class.

Could you introduce a `flushOnFail` option in the 2 mentioned properties with a default value of `true` ? Then use the `$this->getOption('flushOnFail')` to run the condition inside the `handleRequest()` method.

File: src/Forms/Form.php [hunk 2]
```diff
@@ -299,7 +306,9 @@ public function handleRequest(Request $request): FormInterface
             $this->getFormPlaceholders($fields),
         );
 
-        $data = $this->validator->valid();
+        $data = $this->validatorReturnValueOnFail
+            ? $this->validator->getData()
+            : $this->validator->valid();
```
Comment (jlambe) [inline src/Forms/Form.php:311]:
```suggestion
        $data = $this->getOption('flushOnFail', true)
            ? $this->validator->valid()
            : $this->validator->getData();
```

File: src/Forms/Form.php [hunk 3]
```diff
@@ -402,6 +411,21 @@ protected function getFormPlaceholders(array $fields)
...
+     *
+     * @return $this
+     */
+    public function setValidatorReturnValueOnFail($value = true)
+    {
+        $this->validatorReturnValueOnFail = $value;
+
+        return $this;
+    }
+
```
Comment (jlambe) [inline src/Forms/Form.php:428]:
```suggestion
```

You can delete this method. Make sure to add the `flushOnFail` option key to Form `$allowedOptions` property and add `'flushOnFail' => true` to the `$defaultOptions` property.

File: tests/Forms/FormCreationTest.php [hunk 4]
```diff
@@ -1301,6 +1301,54 @@ public function test_checkbox_field_type_with_default_value_can_be_overriden()
...
+        $fields = $this->getFieldsFactory();
+        $field = $fields->text('firstname', [
+            'rules' => 'required|max:3',
+        ]);
+
+        $form = $factory->make()
+            ->add($field)
+            ->get();
+
+        $form->setValidatorReturnValueOnFail();
```
Comment (jlambe) [inline tests/Forms/FormCreationTest.php:1340]:
```suggestion
        $form = $factory->make(null, [
            'flushOnFail' => false,
        ])
            ->add($field)
            ->get();
```

**Comment by Juje** [comment]
The option API actually makes a bit more sense in this case so I've updated everything and the test passes :slightly_smiling_face:.

**Review by jlambe — APPROVED** [review]
Great job @Juje! Thanks for this update.

**Comment by jlambe** [comment]
@Juje I just saw the PR was on the `main` branch. In the future if you have new code requests, target the `3.0` branch instead 😃

**Comment by Juje** [comment]
> @Juje I just saw the PR was on the `main` branch. In the future if you have new code requests, target the `3.0` branch instead smiley

Thanks for accepting the PR :slightly_smiling_face: .

But uhm you told me to merge it against `main` :wink: https://github.com/themosis/framework/issues/874#issuecomment-1289259191

**Comment by jlambe** [comment]
🤦‍♂️ oups 😂

**Comment by Juje** [comment]
But, I updated composer and the change works perfectly with our code on staging :slightly_smiling_face:.

**Pull request merged**